src/channel-qvm.lisp

```diff
@@ -0,0 +1,207 @@
+    ;;
+    ;; [1]: https://en.wikipedia.org/wiki/Inverse_transform_sampling
+    (loop :for kj :in kraus-ops
+          :do
+             (replace amps (amplitudes qvm))
+             (apply-matrix-operator (magicl-matrix-to-quantum-operator kj)
+                                    amps
+                                    (apply #'nat-tuple qubits))
+             (incf summed (psum #'probability amps))
+          :until (>= summed r))
+    (rotatef (amplitudes qvm) (%trial-amplitudes qvm))
+    (normalize-wavefunction (amplitudes qvm))))
+
+
+(defmethod apply-classical-readout-noise ((qvm channel-qvm) (instr quil:measure-discard))
+  (declare (ignore qvm instr))
+  nil)
+
+
+(defmethod apply-classical-readout-noise ((qvm channel-qvm) (instr quil:measure))
+  (%corrupt-readout qvm instr (readout-povms (noise-model qvm))))
+
+
+(defmethod apply-classical-readout-noise ((qvm channel-qvm) (instr compiled-measurement))
+  (apply-classical-readout-noise qvm (source-instruction instr)))
+
+
+(defun %corrupt-readout (qvm instr povm-map)
+  "Apply POVM to measurement."
+  (check-type instr quil:measure)
+  (let* ((q (quil:qubit-index (quil:measurement-qubit instr)))
+         (a (quil:measure-address instr))
+         (c (dereference-mref qvm a))
+         (povm (gethash q povm-map)))
+    (when povm
+      (destructuring-bind (p00 p01 p10 p11) povm
+        (setf (dereference-mref qvm a)
+              (perturb-measurement c p00 p01 p10 p11))))))
+
+
+;;; Don't compile things for the CHANNEL-QVM.
+(defmethod compile-loaded-program ((qvm channel-qvm))
+  qvm)
+
+
+(defmethod compile-instruction ((qvm channel-qvm) isn)
+  (declare (ignore qvm))
+  isn)
+
+
+;; Duplicate fn in noisy-qvm.lisp (deleted in noisy-qvm.lisp)
+(defun perturb-measurement (actual-outcome p00 p01 p10 p11)
+  "Given the readout error encoded in the POVM (see documentation of NOISY-QVM), randomly sample the observed (potentially corrupted) measurement outcome."
+  (check-type actual-outcome bit)
+  (check-type p00 (double-float 0.0d0 1.0d0))
+  (check-type p01 (double-float 0.0d0 1.0d0))
+  (check-type p10 (double-float 0.0d0 1.0d0))
+  (check-type p11 (double-float 0.0d0 1.0d0))
+  (let ((r (random 1.0d0)))
+    (ecase actual-outcome
+      ((0) (if (<= r p00) 0 1))
+      ((1) (if (<= r p01) 0 1)))))
+
+
+;; Duplicate fn in noisy-qvm.lisp (deleted in noisy-qvm.lisp)
+(defun check-povm (povm)
+  "Verify that the list POVM contains a valid single qubit diagonal POVM. Also see the documentation for the READOUT-POVMS slot of NOISY-QVM."
+  (destructuring-bind (p00 p01 p10 p11) povm
+    (check-type p00 (double-float 0.0d0 1.0d0))
+    (check-type p01 (double-float 0.0d0 1.0d0))
+    (check-type p10 (double-float 0.0d0 1.0d0))
+    (check-type p11 (double-float 0.0d0 1.0d0))
+    (assert (cl-quil::double= 1.0d0 (+ p00 p10)))
+    (assert (cl-quil::double= 1.0d0 (+ p01 p11)))))
+
+
+(defun noisy-program-strings (parsed-program qvm)
```

Comment:
im dubious of the value of returning a bunch of strings

Reply:
do you think this would be better as a function that is exclusively used for testing? Like should I move this to a test file?

Reply:
yes